### PR TITLE
Issue #4074: fix overlap between lf and crlf at file end

### PR DESCRIPTION
**Pull request by strkkk**
#4074

Report https://strkkk.github.io/checkstyle/4074/index.html
Report v2 https://strkkk.github.io/checkstyle/4074_2/index.html

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/NewlineAtEndOfFileCheckTest.java [hunk 1]
```diff
@@ -171,7 +185,7 @@ public void testEmptyFileFile() throws Exception {
     public void testFileWithEmptyLineOnly() throws Exception {
         final DefaultConfiguration checkConfig =
                 createModuleConfig(NewlineAtEndOfFileCheck.class);
-        checkConfig.addAttribute("lineSeparator", LineSeparatorOption.LF.toString());
+        checkConfig.addAttribute("lineSeparator", LineSeparatorOption.CRLF.toString());
```
Comment (strkkk) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/NewlineAtEndOfFileCheckTest.java]: The actual separator in a file is CRLF, so this change is needed.

**Comment by strkkk** [comment]
> [ERROR]   NewlineAtEndOfFileCheckTest.testNewlineLfAtEndOfFileLfNotOverlapWithCrLf:74->AbstractModuleTestSupport.verify:236->AbstractModuleTestSupport.verify:256->AbstractModuleTestSupport.verify:292 error message 0 expected:<...dOfFileCrLf.java:1: [File does not end with a newline.]> but was:<...dOfFileCrLf.java:1: [Got an exception - /home/travis/build/checkstyle/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/newlineatendoffile/InputNewlineAtEndOfFileCrLf.java (No such file or directory)]>

Very weird, cannot find file on travis and pitest envs.
Works on AppVeyor and my local machine (win10), and file was there (I reused existing one in repo).

Any ideas?

**Comment by pbludov** [comment]
> Any ideas?

There is no such file `InputNewlineAtEndOfFileCrLf.java` in the folder `src/test/resources/com/puppycrawl/tools/checkstyle/checks/newlineatendoffile/`

There is a file with similar name: `InputNewlineAtEndOfFileCrlf.java` (note lowercase `lf`).

**Comment by strkkk** [comment]
Indeed.

And windows is ok with it since file names are caseless.

Thank you! :)

**Comment by rnveach** [comment]
CI failure in wercker on Spring Integration project.
> [ant:checkstyle] [ERROR] /pipeline/source/.ci-temp/spring-integration/spring-integration-jdbc/src/test/java/org/springframework/integration/jdbc/LockInterceptor.java:1: File does not end with a newline. [NewlineAtEndOfFile]

Issue was started with project. A PR will need to be submitted and accepted to have CI pass.

**Comment by strkkk** [comment]
issue in spring-integration is fixed, I restarted wercker

**Comment by romani** [comment]
https://strkkk.github.io/checkstyle/4074/apache-ant/index.html#A1

>File does not end with a newline.

With new logic, message become weird, as file end with new line, user will not see a reason of problem easily.
Looks like we need to change wording to something like `Detected line end is ... Expected line end is ...`

**Comment by romani** [comment]
@pbludov, @rnveach please share your thoughts on this.

**Comment by rnveach** [comment]
> File does not end with the expected line ending.

Or something similar.

**Comment by romani** [comment]
I would print expected value in message, as some users do not even know where their checkstyle config is located and how to match message to name of check and ... .

**Comment by strkkk** [comment]
I would add abbreviations as well.
As for me, something like
`Expected line ending for file is LF(\n), but CRLF(\r\n) is detected.`
sounds better.

**Comment by rnveach** [comment]
I am fine with @romani s with the extra text.

**Comment by romani** [comment]
@strkkk , looks like majority is in favor of message change - please update message and regenerate report.

**Comment by strkkk** [comment]
@romani you mean message that you suggested? What about adding
abbreviations there (LF and CRLF)?

пн, 27 янв. 2020 г., 11:49 Roman Ivanov <notifications@github.com>:

**Comment by romani** [comment]
> What about adding abbreviations there (LF and CRLF)?

I though you noticed that I supported your idea by :+1: .
I do like your version of message a lot ! It will be very good for users.

**Comment by strkkk** [comment]
> please update message and regenerate report.

done, report link in first post

**Pull request merged**